sdk/formrecognizer/azure-ai-formrecognizer/azure/ai/formrecognizer/_custom_form_client.py

```diff
@@ -0,0 +1,362 @@
+
+
+class CustomFormClient(object):
+    """CustomFormClient.
+
+    :param str endpoint: Supported Cognitive Services endpoints (protocol and hostname,
+        for example: https://westus2.api.cognitive.microsoft.com).
+    :param credential: Credentials needed for the client to connect to Azure.
+        This is an instance of FormRecognizerApiKeyCredential if using an API key.
+    :type credential: ~azure.ai.formrecognizer.FormRecognizerApiKeyCredential
+    """
+
+    def __init__(self, endpoint, credential, **kwargs):
+        # type: (str, FormRecognizerApiKeyCredential, Any) -> None
+        self._client = FormRecognizer(
+            endpoint=endpoint,
+            credential=credential,
+            authentication_policy=CognitiveServicesCredentialPolicy(credential),  # TODO: replace with core policy
+            **kwargs
+        )
+
+    @distributed_trace
+    def begin_training(self, source, source_prefix_filter="", include_sub_folders=False, **kwargs):
+        # type: (str, Optional[str], Optional[bool], Any) -> LROPoller
+        """Create and train a custom model. The request must include a source parameter that is an
+        externally accessible Azure storage blob container Uri (preferably a Shared Access Signature Uri).
+        Models are trained using documents that are of the following content type - 'application/pdf',
+        'image/jpeg', 'image/png', 'image/tiff'. Other type of content is ignored.
+
+        :param str source: An Azure Storage blob container URI.
+        :param str source_prefix_filter: A case-sensitive prefix string to filter documents in the source path for
+            training. For example, when using a Azure storage blob Uri, use the prefix to restrict sub
+            folders for training.
+        :param bool include_sub_folders: A flag to indicate if sub folders within the set of prefix folders
+            will also need to be included when searching for content to be preprocessed.
+        :return: LROPoller
+        :rtype: ~azure.core.polling.LROPoller
+        :raises: ~azure.core.exceptions.HttpResponseError
+        """
+
+        cls = kwargs.pop("cls", None)
+        response = self._client.train_custom_model_async(
+            train_request=TrainRequest(
+                source=source,
+                source_filter=TrainSourceFilter(
+                    prefix=source_prefix_filter,
+                    include_sub_folders=include_sub_folders
+                )
+            ),
+            cls=lambda pipeline_response, _, response_headers: pipeline_response,
+            **kwargs
+        )
+
+        def callback(raw_response):
+            model = self._client._deserialize(Model, raw_response)
+            return CustomModel._from_generated(model)
+
+        # FIXME: https://github.com/Azure/azure-sdk-for-python/issues/10417
+        response.http_response.headers["Location"] = response.http_response.headers["Location"] + "?includeKeys=true"
```

Comment:
This relies on the fact that the (opaque) location url has no query parameters. Which is a fragile assumption.

Reply:
Yeah not happy about this, unfortunately without it we don't get the extracted fields from the training. I can remove for now until there's a better solution - the user can always call `get_custom_model` with the model ID to see the fields.

Reply:
Fixed this with a custom polling method